File: apps/mobile/test/picking_test.dart
// A19 en la app: el gasto nace sabiendo a quién espera, cambiar tu consumo
// te devuelve a «eligiendo» en el mismo commit, y «He terminado» te saca.
import 'package:domain/domain.dart';
import 'package:fake_cloud_firestore/fake_cloud_firestore.dart';
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:salda_mobile/features/sessions/data/firestore_session_repository.dart';
import 'package:salda_mobile/features/sessions/domain/session_models.dart';
import 'package:salda_mobile/features/sessions/presentation/ticket_detail_screen.dart';
import 'package:salda_mobile/l10n/generated/app_localizations.dart';

import 'fakes.dart';

const _ticketPath = 'sessions/s1/accounts/a1/tickets/t1';
const _linePath = '$_ticketPath/lines/l1';

FirestoreSessionRepository _repo(FakeFirebaseFirestore fake) =>
    FirestoreSessionRepository(
      firestore: fake,
      uid: () => 'owner',
      shareCodeFactory: () => 'TEST-CODE-1234567890',
    );

/// Ticket bajo el protocolo, con p1 (dueña) y p2 (Jorge) pendientes.
Future<FakeFirebaseFirestore> _seed({bool protocolo = true}) async {
  final fake = FakeFirebaseFirestore();
  await fake.doc('sessions/s1').set({
    'ownerUid': 'owner',
    'kind': 'single',
    'name': 'Cena',
    'status': 'open',
    'splitModeDefault': 'byItem',
    'shareCode': 'TEST-CODE-1234567890',
    'currency': 'EUR',
  });
  await fake.doc('sessions/s1/participants/p1').set({
    'name': 'Edgar',
    'isOwner': true,
    'order': 0,
    'claimedByDevice': 'owner',
    'active': true,
  });
  await fake.doc('sessions/s1/participants/p2').set({
    'name': 'Alba',
    'isOwner': false,
    'order': 1,
    'claimedByDevice': 'dev-2',
    'active': true,
  });
  await fake.doc(_ticketPath).set({
    'kind': 'manual',
    'grandTotal': 400,
    'paidByParticipantId': 'p1',
    if (protocolo) 'pickingModelVersion': 1,
    if (protocolo)
      'picking': {
        'open': {'p1': true, 'p2': true},
      },
  });
  await fake.doc(_linePath).set({
    'name': 'Flauta',
    'order': 0,
    'quantityMilli': 2000,
    'totalPrice': 400,
    'unitIds': ['u0', 'u1'],
    'assignment': {
      'type': 'units',
      'schemaVersion': 2,
      'units': <String, Object?>{},
    },
  });
  return fake;
}

void main() {
  group('A19: siembra al crear el gasto', () {
    test('un ticket nuevo nace con el protocolo y todos pendientes', () async {
      final repo = _repo(FakeFirebaseFirestore());
      final creada = await repo.createSession(
        NewSessionInput(
          name: 'Cena',
          kind: 'single',
          participantNames: const ['Edgar', 'Alba', 'Tete'],
          payerIndex: 0,
          splitModeDefault: SplitMode.byItem,
          ticket: const NewTicketInput(
            merchantName: 'Bar',
            grandTotal: Money(600),
            kind: 'manual',
            engine: 'manual',
            confidence: 1.0,
            lines: [NewLineInput(name: 'Cerveza', totalPrice: Money(600))],
          ),
        ),
      );
      final ticket = await repo.firestore.doc(creada.ticketPath).get();
      expect(ticket.data()!['pickingModelVersion'], 1);
      expect((ticket.data()!['picking'] as Map)['open'], {
        'p0': true,
        'p1': true,
        'p2': true,
      });
    });

    test(
      'añadir un gasto siembra SOLO con los participantes activos',
      () async {
        final repo = _repo(FakeFirebaseFirestore());
        final creada = await repo.createSession(
          NewSessionInput(
            name: 'Cena',
            kind: 'single',
            participantNames: const ['Edgar', 'Alba'],
            payerIndex: 0,
            splitModeDefault: SplitMode.byItem,
            spaceId: 'gr1',
            ticket: const NewTicketInput(
              merchantName: 'Bar',
              grandTotal: Money(600),
              kind: 'manual',
              engine: 'manual',
              confidence: 1.0,
              lines: [NewLineInput(name: 'Cerveza', totalPrice: Money(600))],
            ),
          ),
        );
        // Alba deja de participar antes del segundo gasto.
        await repo.firestore
            .doc('sessions/${creada.sessionId}/participants/p1')
            .update({'active': false});

        final path = await repo.addTicket(
          creada.sessionId,
          const NewTicketInput(
            merchantName: 'Otro',
            grandTotal: Money(200),
            kind: 'manual',
            engine: 'manual',
            confidence: 1.0,
            lines: [NewLineInput(name: 'Café', totalPrice: Money(200))],
          ),
          payerPid: 'p0',
          spaceId: 'gr1',
        );
        final ticket = await repo.firestore.doc(path).get();
        // A quien ya no participa no se le espera: recompute lo descarta igual.
        expect((ticket.data()!['picking'] as Map)['open'], {'p0': true});
      },
    );
  });

  group('A19: escrituras del protocolo', () {
    test('elegir consumo reabre a esa persona en el MISMO commit', () async {
      final fake = await _seed();
      final repo = _repo(fake);
      await repo.finishPicking(_ticketPath, participantId: 'p2');
      expect(
        ((await fake.doc(_ticketPath).get()).data()!['picking'] as Map)['open'],
        {'p1': true},
      );

      await repo.setUnitConsumer(
        _linePath,
        unit: 0,
        participantId: 'p2',
        selected: true,
        myPid: 'p2',
        usesPicking: true,
      );
      final picking =
          (await fake.doc(_ticketPath).get()).data()!['picking'] as Map;
      expect((picking['open'] as Map)['p2'], true);
      // El objetivo se declara: es lo que verifican las Rules.
      expect(picking['lastTarget'], 'p2');
    });

    test('en un gasto anterior al protocolo NO se escribe picking', () async {
      final fake = await _seed(protocolo: false);
      final repo = _repo(fake);
      await repo.setUnitConsumer(
        _linePath,
        unit: 0,
        participantId: 'p2',
        selected: true,
        myPid: 'p2',
      );
      // Escribirle `picking` haría que las Rules rechazasen el lote entero.
      expect(
        (await fake.doc(_ticketPath).get()).data()!.containsKey('picking'),
        isFalse,
      );
    });

    test('He terminado saca mi pid de la lista de pendientes', () async {
      final fake = await _seed();
      final repo = _repo(fake);
      await repo.finishPicking(_ticketPath, participantId: 'p1');
      final picking =
          (await fake.doc(_ticketPath).get()).data()!['picking'] as Map;
      expect((picking['open'] as Map).containsKey('p1'), isFalse);
      expect(picking['lastTarget'], 'p1');
    });

    test('el ticket transporta quién falta por terminar', () async {
      final fake = await _seed();
      final tickets = await _repo(fake).watchTickets('s1', 'a1').first;
      expect(tickets.single.usesPicking, isTrue);
      expect(tickets.single.pickingOpen, {'p1', 'p2'});
    });
  });

  group('A19: la pantalla dice a quién se espera', () {
    Future<void> pump(WidgetTester tester, FakeFirebaseFirestore fake) async {
      tester.view.physicalSize = const Size(800, 2000);
      tester.view.devicePixelRatio = 1.0;
      addTearDown(tester.view.reset);
      await tester.pumpWidget(
        ProviderScope(
          overrides: loggedInOverrides(firestore: fake),
          child: MaterialApp(
            localizationsDelegates: AppLocalizations.localizationsDelegates,
            supportedLocales: AppLocalizations.supportedLocales,
            home: TicketDetailScreen(
              ticket: TicketRef(
                sessionId: 's1',
                payerName: 'Edgar',
                ticket: SessionTicket(
                  id: 't1',
                  path: _ticketPath,
                  merchantName: 'Bar Manolo',
                  grandTotal: const Money(400),
                  paidBy: 'p1',
                  kind: 'manual',
                  splitModeOverride: SplitMode.byItem,
                  pickingModelVersion: 1,
                  pickingOpen: const {'p1', 'p2'},
                ),
              ),
            ),
          ),
        ),
      );
      await tester.pumpAndSettle();
    }

    testWidgets('un gasto abierto enumera quién falta y ofrece He terminado', (
      tester,
    ) async {
      await pump(tester, await _seed());
      expect(find.text('Aún estáis eligiendo'), findsOneWidget);
      expect(find.textContaining('Falta Edgar, Alba'), findsOneWidget);
      expect(find.text('He terminado'), findsOneWidget);
    });

    testWidgets('con autoridad A10 se puede terminar por otra persona', (
      tester,
    ) async {
      await pump(tester, await _seed());
      // El dueño de la sesión tiene A10: cierra por Alba.
      expect(find.text('Terminar por Alba'), findsOneWidget);
    });
  });
}

File: backend/firestore/test/picking.test.mjs
/**
 * A19 — cierre de consumo: quién puede terminar, quién puede reabrir, y por
 * qué es IMPOSIBLE dejar un «he terminado» obsoleto.
 *
 * El invariante que fijan estas pruebas: en un ticket bajo el protocolo,
 * ninguna escritura de reparto pasa si el pid afectado no queda abierto tras
 * el commit. Y una regresión de presupuesto: el camino de A10 sobre un
 * MANUAL es el que agotaba las 1000 expresiones cuando la comprobación
 * costaba DOS accesos al ticket; con uno cabe.
 *
 * Ejecutar desde la raíz del repo:
 *   firebase emulators:exec --only firestore,storage --project demo-salda \
 *     "npm --prefix backend/firestore test"
 */
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { after, before, beforeEach, describe, it } from 'node:test';
import {
  assertFails,
  assertSucceeds,
  initializeTestEnvironment,
} from '@firebase/rules-unit-testing';
import {
  deleteField,
  doc,
  serverTimestamp,
  setDoc,
  updateDoc,
  writeBatch,
} from 'firebase/firestore';

let env;

const ALBA = 'uid-alba'; // dueña de la sesión
const JEFA = 'uid-jefa'; // propietaria del grupo (autoridad A10)
const JORGE = 'uid-jorge'; // miembro normal, participante p2
const EDGAR = 'uid-edgar'; // miembro normal, participante p6
const INVITADA = 'uid-invitada'; // anónima con guestAccess, reclama p5

const T = 'sessions/sg1/accounts/a1/tickets/t1';
const L = `${T}/lines/l1`;
const UNIDADES = 6;

const db = (uid) =>
  env
    .authenticatedContext(
      uid,
      uid === INVITADA
        ? { firebase: { sign_in_provider: 'anonymous' } }
        : {
            email: `${uid}@salda.test`,
            email_verified: true,
            firebase: { sign_in_provider: 'password' },
          },
    )
    .firestore();

/** Elegir consumo: línea + reapertura del pid, en UN batch de dos docs. */
const elegir = (
  actor,
  pid,
  { unit = 'u0', selected = true, firma, reabre = true } = {},
) => {
  const f = db(actor);
  const b = writeBatch(f);
  b.update(doc(f, L), {
    'assignment.type': 'units',
    'assignment.schemaVersion': 2,
    'assignment.lastEditorPid': pid,
    'assignment.lastEditedUnit': unit,
    [`assignment.units.${unit}.${pid}`]: selected ? true : deleteField(),
    [`assignment.by.${unit}.${pid}`]:
      selected && firma ? firma : deleteField(),
  });
  if (reabre) {
    b.update(doc(f, T), {
      'picking.lastTarget': pid,
      [`picking.open.${pid}`]: true,
    });
  }
  return b.commit();
};

/** «He terminado»: el pid sale del mapa de pendientes. */
const terminar = (actor, pid) =>
  updateDoc(doc(db(actor), T), {
    'picking.lastTarget': pid,
    [`picking.open.${pid}`]: deleteField(),
  });

/** Deniega, y deniega DE VERDAD: no por quedarse sin presupuesto. */
async function assertDeniegaLimpio(promise) {
  await assert.rejects(promise, (error) => {
    const mensaje = `${error?.message ?? error}`;
    assert.ok(
      !mensaje.includes('maximum of'),
      `Rules se quedó sin presupuesto en vez de denegar: ${mensaje}`,
    );
    return true;
  });
}

before(async () => {
  env = await initializeTestEnvironment({
    projectId: 'demo-salda',
    firestore: {
      rules: readFileSync(new URL('../firestore.rules', import.meta.url), 'utf8'),
    },
  });
});
after(async () => env?.cleanup());

/** `protocolo: false` siembra un ticket anterior a A19. */
async function sembrar({ protocolo = true, unidades = UNIDADES } = {}) {
  await env.clearFirestore();
  await env.withSecurityRulesDisabled(async (ctx) => {
    const f = ctx.firestore();
    for (const uid of [ALBA, JEFA, JORGE, EDGAR]) {
      await setDoc(doc(f, `profiles/${uid}`), {
        displayName: uid, displayNameLower: uid, username: uid,
        createdAt: serverTimestamp(), updatedAt: serverTimestamp(),
        schemaVersion: 1,
      });
    }
    await setDoc(doc(f, 'spaces/gr1'), {
      name: 'Piso', ownerUid: JEFA, kind: 'group', status: 'active',
      createdAt: serverTimestamp(), schemaVersion: 2,
    });
    for (const uid of [JEFA, ALBA, JORGE, EDGAR]) {
      await setDoc(doc(f, `spaces/gr1/members/${uid}`), {
        uid, joinedAt: serverTimestamp(),
      });
    }
    await setDoc(doc(f, 'sessions/sg1'), {
      ownerUid: ALBA, kind: 'single', name: 'Cena', status: 'open',
      splitModeDefault: 'byItem', shareCode: 'SECRET-CODE-16CHARS',
      currency: 'EUR', contextModelVersion: 1, spaceId: 'gr1',
      computeVersion: 0, totals: {}, balances: {},
    });
    await setDoc(doc(f, 'sessions/sg1/participants/p1'), {
      name: 'Alba', isOwner: true, order: 0, active: true, claimedByDevice: ALBA,
    });
    await setDoc(doc(f, 'sessions/sg1/participants/p2'), {
      name: 'Jorge', isOwner: false, order: 1, active: true,
      claimedByDevice: JORGE,
    });
    // Persona MANUAL: no reclama nada y aun así consume. Solo A10 la cierra.
    await setDoc(doc(f, 'sessions/sg1/participants/p3'), {
      name: 'Tete', isOwner: false, order: 2, active: true,
      claimedByDevice: '', manualId: 'm-tete',
    });
    await setDoc(doc(f, 'sessions/sg1/participants/p5'), {
      name: 'Invitada', isOwner: false, order: 4, active: true,
      claimedByDevice: INVITADA,
    });
    await setDoc(doc(f, 'sessions/sg1/participants/p6'), {
      name: 'Edgar', isOwner: false, order: 5, active: true,
      claimedByDevice: EDGAR,
    });
    await setDoc(doc(f, `sessions/sg1/guestAccess/${INVITADA}`), {
      shareCode: 'SECRET-CODE-16CHARS',
    });
    await setDoc(doc(f, 'sessions/sg1/accounts/a1'), { name: 'Cena' });
    await setDoc(doc(f, T), {
      kind: 'manual', grandTotal: 6000, paidByParticipantId: 'p1',
      merchant: { name: 'Bar' }, spaceId: 'gr1', contextModelVersion: 1,
      splitModeOverride: 'byItem',
      ...(protocolo
        ? {
            pickingModelVersion: 1,
            picking: {
              open: { p1: true, p2: true, p3: true, p5: true, p6: true },
            },
          }
        : {}),
    });
    await setDoc(doc(f, L), {
      name: 'Cerveza', totalPrice: 6000, quantityMilli: unidades * 1000,
      order: 0,
      unitIds: Array.from({ length: unidades }, (_, i) => `u${i}`),
      assignment: { type: 'units', schemaVersion: 2, units: {} },
    });
  });
}

describe('A19: elegir con el protocolo activo', () => {
  beforeEach(() => sembrar());

  it('un participante activo elige lo suyo', () =>
    assertSucceeds(elegir(JORGE, 'p2')));

  it('la autoselección NO lleva firma de tercero', () =>
    assertSucceeds(elegir(JORGE, 'p2', { firma: undefined })));

  it('A10 asigna a otra persona CON procedencia', () =>
    assertSucceeds(elegir(JEFA, 'p2', { firma: JEFA })));

  it('A10 sin firma sobre un tercero queda denegado', () =>
    assertDeniegaLimpio(elegir(JEFA, 'p3', { firma: undefined })));

  it('un participante DESACTIVADO no elige', async () => {
    await env.withSecurityRulesDisabled(async (ctx) => {
      await updateDoc(doc(ctx.firestore(), 'sessions/sg1/participants/p2'), {
        active: false,
      });
    });
    await assertDeniegaLimpio(elegir(JORGE, 'p2'));
  });

  it('un ex-miembro del grupo no elige', async () => {
    await env.withSecurityRulesDisabled(async (ctx) => {
      const { deleteDoc } = await import('firebase/firestore');
      await deleteDoc(doc(ctx.firestore(), `spaces/gr1/members/${JORGE}`));
    });
    await assertDeniegaLimpio(elegir(JORGE, 'p2'));
  });

  it('con la sesión cerrada no elige nadie', async () => {
    await env.withSecurityRulesDisabled(async (ctx) => {
      await updateDoc(doc(ctx.firestore(), 'sessions/sg1'), {
        status: 'closed',
      });
    });
    await assertDeniegaLimpio(elegir(JORGE, 'p2'));
  });

  it('la invitada web sigue exactamente el mismo protocolo', async () => {
    await assertSucceeds(elegir(INVITADA, 'p5', { unit: 'u3' }));
    await assertSucceeds(terminar(INVITADA, 'p5'));
    await assertDeniegaLimpio(
      elegir(INVITADA, 'p5', { unit: 'u4', reabre: false }),
    );
  });
});

describe('A19: la reapertura es obligatoria y atómica', () => {
  beforeEach(() => sembrar());

  it('quien ya terminó NO puede cambiar sin reabrirse', async () => {
    await terminar(JORGE, 'p2');
    await assertDeniegaLimpio(
      elegir(JORGE, 'p2', { unit: 'u1', reabre: false }),
    );
  });

  it('cambio y reapertura juntos, en el mismo commit, se permiten', async () => {
    await terminar(JORGE, 'p2');
    await assertSucceeds(elegir(JORGE, 'p2', { unit: 'u1' }));
  });

  it('A10 cambia el consumo de un MANUAL y lo reabre en el mismo commit',
    async () => {
      await terminar(JEFA, 'p3');
      await assertDeniegaLimpio(
        elegir(JEFA, 'p3', { unit: 'u2', firma: JEFA, reabre: false }),
      );
      // REGRESIÓN DE PRESUPUESTO: este es el camino que agotaba las 1000
      // expresiones cuando la comprobación costaba dos accesos al ticket.
      await assertSucceeds(elegir(JEFA, 'p3', { unit: 'u2', firma: JEFA }));
    });

  it('reabrir a OTRA persona sin autoridad queda denegado', async () => {
    await terminar(JORGE, 'p2');
    // Edgar intenta reabrir a Jorge para tocarle el consumo.
    await assertDeniegaLimpio(elegir(EDGAR, 'p2', { firma: EDGAR }));
  });
});

describe('A19: terminar', () => {
  beforeEach(() => sembrar());

  it('cada cual cierra SU pid', () => assertSucceeds(terminar(JORGE, 'p2')));

  it('A10 cierra el pid de otra persona y el de un manual', async () => {
    await assertSucceeds(terminar(JEFA, 'p2'));
    await assertSucceeds(terminar(ALBA, 'p3'));
  });

  it('un participante normal NO cierra el pid de otra persona', () =>
    assertDeniegaLimpio(terminar(JORGE, 'p1')));

  it('la invitada cierra el suyo y no el ajeno', async () => {
    await assertSucceeds(terminar(INVITADA, 'p5'));
    await assertDeniegaLimpio(terminar(INVITADA, 'p6'));
  });

  it('nadie toca la huella de topología', () =>
    assertDeniegaLimpio(
      updateDoc(doc(db(ALBA), T), {
        'picking.lastTarget': 'p1',
        'picking.fingerprint': 'inventada',
      }),
    ));

  it('nadie toca la economía congelada, ni siquiera el dueño', () =>
    assertDeniegaLimpio(
      updateDoc(doc(db(ALBA), T), {
        'picking.lastTarget': 'p1',
        'picking.firmContribution': {
          paidBy: 'p1', grandTotal: 6000, consumption: { p2: 6000 },
        },
      }),
    ));

  it('una escritura de picking no cuela otro campo del ticket', () =>
    assertDeniegaLimpio(
      updateDoc(doc(db(ALBA), T), {
        'picking.lastTarget': 'p2',
        'picking.open.p2': deleteField(),
        grandTotal: 1,
      }),
    ));

  it('no se cierran DOS personas en la misma escritura', () =>
    assertDeniegaLimpio(
      updateDoc(doc(db(ALBA), T), {
        'picking.lastTarget': 'p2',
        'picking.open.p2': deleteField(),
        'picking.open.p6': deleteField(),
      }),
    ));

  it('terminar sigue permitido con la economía ya congelada', async () => {
    await env.withSecurityRulesDisabled(async (ctx) => {
      await updateDoc(doc(ctx.firestore(), T), {
        'picking.firmContribution': {
          paidBy: 'p1', grandTotal: 6000, consumption: { p2: 6000 },
        },
      });
    });
    await assertSucceeds(terminar(JORGE, 'p2'));
  });
});

describe('A19: compatibilidad y presupuesto', () => {
  it('un ticket SIN pickingModelVersion se comporta como hoy', async () => {
    await sembrar({ protocolo: false });
    await assertSucceeds(elegir(JORGE, 'p2', { reabre: false }));
  });

  it('cliente antiguo con el pid abierto sigue pudiendo editar', async () => {
    await sembrar();
    await assertSucceeds(elegir(JORGE, 'p2', { reabre: false }));
  });

  it('cliente antiguo con el pid cerrado queda denegado', async () => {
    await sembrar();
    await terminar(JORGE, 'p2');
    await assertDeniegaLimpio(
      elegir(JORGE, 'p2', { unit: 'u2', reabre: false }),
    );
  });

  it('el dueño crea el ticket con el protocolo ya sembrado', async () => {
    await sembrar();
    await assertSucceeds(
      setDoc(doc(db(ALBA), 'sessions/sg1/accounts/a1/tickets/t9'), {
        kind: 'manual', grandTotal: 1000, paidByParticipantId: 'p1',
        merchant: { name: 'Bar' }, spaceId: 'gr1', contextModelVersion: 1,
        splitModeOverride: 'byItem',
        pickingModelVersion: 1,
        picking: { open: { p1: true, p2: true } },
      }),
    );
  });

  it('el dueño NO puede vaciar `open` por la puerta de siempre', async () => {
    await sembrar();
    await assertDeniegaLimpio(
      updateDoc(doc(db(ALBA), T), {
        grandTotal: 6000,
        paidByParticipantId: 'p1',
        picking: { open: {} },
      }),
    );
  });

  it('presupuesto: 24 unidades seguidas con la comprobación extra',
    async () => {
      await sembrar({ unidades: 24 });
      for (let u = 0; u < 24; u++) {
        await assertSucceeds(elegir(JORGE, 'p2', { unit: `u${u}` }));
      }
    });
});
